### PR TITLE
feat(ec2): support reserving azs space in VPCs

### PR DESCRIPTION
**Pull request by yogeshchopra**
VPCProps take a new argument `reservedAzs` which can be used to allocate IP space for availability zones in the VPC without actually instantiating the VPC resources.

Closes https://github.com/aws/aws-cdk/issues/6683

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/README.md [hunk 1]
```diff
@@ -114,6 +114,31 @@ class MyStack extends Stack {
 
 Note that overriding the `get availabilityZones()` method will override the default behavior for all constructs defined within the Stack.
 
+### Reserving availability zones
+
+There are situations where the IP space for availability zones will
+need to be reserved. This is useful in situations where availability
+zones would need to be added after the vpc is originally deployed,
+without causing IP renumbering for availability zones subnets. The IP
+space for reserving n availability zones can be done by setting the
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ec2/README.md:123]: I think that n is intentional here to mean that you're not sure how many, but the wording may not sound clear to others.

Comment (TheRealAmazonKendra) [reply packages/@aws-cdk/aws-ec2/README.md:123]: Editing to add that now that I've read the line below, I see that it was indeed intentional, but I think the wording is a little clunky here. Maybe at least make n => `n`.

File: packages/@aws-cdk/aws-ec2/README.md [hunk 2]
```diff
@@ -114,6 +114,31 @@ class MyStack extends Stack {
...
+```ts
+const vpc = new ec2.Vpc(this, 'TheVPC', {
+  cidr: '10.0.0.0/21',
+  maxAzs: 3,
+  reservedAzs: 1,
+});
+```
+
+In the example above, the subnets for reserved availability zones is not
+actually provisioned but its IP space is still reserved. If in the future
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ec2/README.md:135]:
```suggestion
actually provisioned but its IP space is still reserved. If, in the future,
```

File: packages/@aws-cdk/aws-ec2/lib/vpc.ts [hunk 3]
```diff
@@ -877,6 +877,16 @@ export interface VpcProps {
...
 
+  /**
+   * Define the number of AZs to reserve.
+   *
+   * When specified, the IP space is reserved for the azs but no actual
+   * resources are provisioned.
+   *
+   * @default 0
+   */
+  readonly reservedAzs?: number;
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/aws-ec2/lib/vpc.ts:888]: I'm not really sure that this is the right fix here. After reading through the comments on the issue, it's clear that we do need a fix here, but how often will people think ahead to reserve additional AZs?

I think our team needs to discuss a path forward on this one so, full disclaimer, this is where I'm stopping this review for now. I don't want to waste your time in edits when I'm just not sure this is the right path forward.

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]

**Comment by yogeshchopra** [comment]
Any update from the team on the same?
This is a blocker for us, as when we want to build vpc we want it to be expandable if needed in future, this will enable to keep our service available in all azs in future.

**Review by rix0rrr — CHANGES_REQUESTED** [review]
Heya! I like the idea here! 

Be aware that we just merged https://github.com/aws/aws-cdk/pull/22458, which changes the IP allocation code substantially. I don't think you necessarily need to change anything about the API, but you may need to move the implementation.

A simple trick that might work for all IP address providers at once might be to just add one or more marker strings to the list of AZ names (`"fake-az"`), let those participate in the calculations as normal, and then not create actual subnet resources for those AZs.

I'd like to see a different form of unit test than the one you currently have though. You have a bunch of numbers in there -- they could be correct, they could be wrong. I'm dumb, and I don't have the inclination to check your math 😬.

But what I'm really interested in is not the CIDR values of some subnets under some conditions. I'm actually more interested in a specific *property*:

> **GIVEN** a Vpc has been created with reserved AZs
> **WHEN** those reserved AZs are replaced with *real* AZs
> **THEN** none of the subnets that already existed would suffer any CIDR changes

This is a test we could be writing, and it doesn't even involve any math! It's a more robust test as well. I don't care about the exact CIDRs for any specific combination of settings -- I just care that whatever the values are, they don't change when I finally go to add that AZ!

Something like (spitballing here):

```ts
const stack1 = new Stack();
const stack2 = net Stack();

new Vpc(stack1, 'Vpc', {
  mazAzs: 2,
  reservedAzs: 1,
});
new Vpc(stack2, 'Vpc', {
  mazAzs: 3,
  reservedAzs: 0,
});

const template1 = Template.fromStack(stack1);
const template2 = Template.fromStack(stack2);

const subnets1 = template1.findResource('AWS::EC2::Subnet');

for (const [logicalId, subnet1] of Object.entries(subnets1)) {
  const subnet2 = template2.template.Resources[logicalId];
  expect(subnet2).toEqual(subnet1);
}
```

Or something, not entirely sure this is 100% but you see what I'm doing here.

Once we have code that tests this *property* (instead of code that tests a *case*), we can even drop a `test.each()` around it to test it for multiple cases, for nearly free:

```ts
test.each([
 [{ mazAzs: 2, reservedAzs: 1 }, {maxAzs: 3 }],
 [{ mazAzs: 2, reservedAzs: 2 }, {maxAzs: 2, reservedAzs: 1 }],
 [/* maybe something with subnetconfig? */],
})('subnets should remain the same going from %p to %p', (props1, props2) => {
  // ....

  new Vpc(stack1, 'Vpc', { ...props1 });
  new Vpc(stack2, 'Vpc', { ...props2 });

  // ...
});
```

**Comment by yogeshchopra** [comment]
How do I reopen this, a merge in my repository closed this pr 😅

**Comment by yogeshchopra** [comment]
@rix0rrr that's really good feedback, will update it, not able to open this pr, will create a new pr incorporating your feedback

**Comment by yogeshchopra** [comment]
https://github.com/aws/aws-cdk/pull/22705

**Pull request closed without merging**